---
title: "agro932 - homework1"
author: "jinlong"
date: "2/19/2020"
output:
  word_document: default
  pdf_document: default
  html_document: default
---


```{r setup}
# , include=TRUE, warning=FALSE, echo=TRUE, error=FALSE
knitr::opts_knit$set(root.dir = normalizePath('../'))
knitr::opts_chunk$set(warning=FALSE, message=FALSE, error=FALSE, echo=TRUE)
```


## Job content
Simulate an NGS dataset for two populations with a small reference genome (< 1 Mb) of yourchoice, each population with 10 diploid individuals, to address a hypothesis about populationdifferentiation.

- Establish a version-controlled directory system to host the project.
- Describe your simulation strategy and the hypothesis to test (positive, negative, or neutral selection).
- Use ANGSD to calculate thetas for each population and Fst between the twopopulations.
- Breakdown the theta ratios and Fst values into different genomic features (i.e., genic and intergenic regions) according to the general feature format (GFF) file for the referencegenome of your choice.
- Visualize and interpret your results and report them in a reproducible manner.

## What I do

### Establish a version-controlled directory system to host the project.
- Establish project **agro932-homework1**: *git@github.com:lijinlong1991/agro932-homework1.git*
- This is [directory system](http://projecttemplate.net/architecture.html)

### Describe your simulation strategy and the hypothesis to test (positive, negative, or neutral selection).
- simulation strategy hypothesis
two population
neutral selection: because the sequences were simulated by same methods. result shows same theta.



#### download the reference genome 
[link](ftp://ftp.ensemblgenomes.org/pub/plants/release-46/fasta/zea_mays/dna/Zea_mays.B73_RefGen_v4.dna.chromosome.Mt.fa.gz)
#### then unzip it
```{bash, eval=FALSE}
wget ftp://ftp.ensemblgenomes.org/pub/plants/release-46/fasta/zea_mays/dna/Zea_mays.B73_RefGen_v4.dna.chromosome.Mt.fa.gz
gunzip Zea_mays.B73_RefGen_v4.dna.chromosome.Mt.fa.gz

```


#### simulate 10 individals
```{bash, eval=FALSE}
for i in {1..10}
do
   wgsim Zea_mays.B73_RefGen_v4.dna.chromosome.Mt.fa -e 0 -d 500 -N 50000 -1 100 -2 100 -r 0.1  -R 0 -X 0 l$i.read1.fq l$i.read2.fq
done
 
```


#### check
```{bash, eval=FALSE}
wc -l l1.read1.fq 

```



#### Align the NGS reads to the reference genome
```{bash, eval=FALSE}
module load bwa samtools
bwa index Zea_mays.B73_RefGen_v4.dna.chromosome.Mt.fa    #index the reference genome # using bwa mem to align the reads to the reference genome
#bwa mem Zea_mays.B73_RefGen_v4.dna.chromosome.Mt.fa l1.read1.fq l1.read2.fq | samtools view -bSh - > l1.bam  # => samtools to convert into bam file

```



#### Do alignment for 2 individuals using bash loop
```{bash, eval=FALSE}
for i in {1..10}; do bwa mem Zea_mays.B73_RefGen_v4.dna.chromosome.Mt.fa l$i.read1.fq l$i.read2.fq | samtools view -bSh - > l$i.bam; done # alignment
for i in *.bam; do samtools sort $i -o sorted_$i; done # sort
for i in sorted*.bam; do samtools index $i; done # index them

```


#### Check mapping statistics
```{bash, eval=FALSE}
samtools flagstat sorted_l1.bam

```



#### write the bam files to a txt file
```{bash, eval=FALSE}
mkdir bam_files
mv sorted*.bam bam_files
cd bam_files
angsd -bam bam.txt -doSaf 1 -anc ../Zea_mays.B73_RefGen_v4.dna.chromosome.Mt.fa -GL 1  -out out 
```


# use realSFS to calculate sfs
```{bash, eval=FALSE}
/home/jyanglab/jli89/bin/angsd/misc/realSFS out.saf.idx > out.sfs
cp out.sfs ../../cache/

angsd -bam bam.txt -out out -doThetas 1 -doSaf 1 -pest out.sfs -anc ../Zea_mays.B73_RefGen_v4.dna.chromosome.Mt.fa -GL 1
/home/jyanglab/jli89/bin/angsd/misc/thetaStat print out.thetas.idx > theta.txt
```


## cp theta to the cache/ folder
```{bash, eval=FALSE}
cp theta.txt ../../cache/
```



#### pop1&pop2
#### first calculate per pop saf for each populatoin
```{bash, eval=FALSE}
angsd -b pop1.txt -anc ../Zea_mays.B73_RefGen_v4.dna.chromosome.Mt.fa -out pop1 -dosaf 1 -gl 1
angsd -b pop2.txt -anc ../Zea_mays.B73_RefGen_v4.dna.chromosome.Mt.fa -out pop2 -dosaf 1 -gl 1
# calculate the 2dsfs prior
/home/jyanglab/jli89/bin/angsd/misc/realSFS pop1.saf.idx pop2.saf.idx > pop1.pop2.ml
# prepare the fst for easy window analysis etc
/home/jyanglab/jli89/bin/angsd/misc/realSFS fst index pop1.saf.idx pop2.saf.idx -sfs pop1.pop2.ml -fstout out
# get the global estimate
/home/jyanglab/jli89/bin/angsd/misc/realSFS fst stats out.fst.idx 
# below is not tested that much, but seems to work
/home/jyanglab/jli89/bin/angsd/misc/realSFS fst stats2 out.fst.idx -win 500 -step 100 > fst_win.txt
```



#### cp sfs to the cache/ folder
cp fst_win.txt ../../cache/


## Breakdown the theta ratios and Fst values into different genomic features (i.e., genic andintergenic regions) according to the general feature format
```{bash, eval=FALSE}
wget ftp://ftp.ensemblgenomes.org/pub/plants/release-46/fasta/zea_mays/dna/Zea_mays.B73_RefGen_v4.dna.chromosome.Mt.fa.gz
gunzip Zea_mays.B73_RefGen_v4.dna.chromosome.Mt.fa.gz

```

---
- then work in the R


## Visualize and interpret your results and report them in a reproducible manner.


```{r}
s <- scan('cache/out.sfs')
s <- s[-c(1,length(s))]
s <- s/sum(s)
barplot(s,names=1:length(s), main='SFS')

t <- read.table("cache/theta.txt")
hist(t[,4])

fst <- read.table("cache/fst_win.txt", skip=1, header=FALSE)
names(fst)[c(3,5)] <- c("midp", "fst")
plot(fst$midp, fst$fst, xlab="Physical position", ylab="Fst", col="#5f9ea0", pch=16)
```



# General feature format (GFF) from EnsemblPlants
```{r}
library("data.table")
## simply read in wouldn't work
#gff <- fread("largedata/Zea_mays.B73_RefGen_v4.46.chromosome.Mt.gff3", skip="#", header=FALSE, data.table=FALSE)
## grep -v means select lines that not matching any of the specified patterns
gff <- fread(cmd='grep -v "#" largedata/Zea_mays.B73_RefGen_v4.46.chromosome.Mt.gff3', header=FALSE, data.table=FALSE)
#head(gff)

names(gff) <- c("seq", "source", "feature", "start", "end", "score", "strand", "phase", "att")
#table(gff$feature)
```
### Get genes and upstream and downstream 5kb regions

```{r}

g <- subset(gff, feature %in% "gene")
g$geneid <- gsub(".*gene:|;biotype.*", "", g$att)


### + strand
gp <- subset(g, strand %in% "+") 
# nrow(gp) 75

### get the 5k upstream of the + strand gene model
gp_up <- gp
gp_up$end <- gp_up$start - 1
gp_up$start <- gp_up$end - 5000 

### get the 5k downstream of the + strand gene model
gp_down <- gp
gp_down$start <- gp_down$end + 1
gp_down$end <- gp_down$start + 5000 
### - strand
gm <- subset(g, strand %in% "-") 
# nrow(gm) 82

### get the 5k upstream of the + strand gene model
gm_up <- gm
gm_up$start <- gm_up$end + 1
gm_up$end <- gm_up$start + 5000 

### get the 5k downstream of the + strand gene model
gm_down <- gm
gm_down$end <- gm_down$start - 1
gm_down$start <- gm_down$end - 5000 

gup <- rbind(gp_up, gm_up)
fwrite(gup, "cache/mt_gene_up5k.txt", sep="\t", row.names = FALSE, quote=FALSE)

gdown <- rbind(gp_down, gm_down)
fwrite(gdown, "cache/mt_gene_down5k.txt", sep="\t", row.names = FALSE, quote=FALSE)
```




## Get genes and upstream and downstream 5kb regions

```{r}
### - strand
gm <- subset(g, strand %in% "-") 
dim(gm) # 82
fwrite(g, "cache/mt_gene.txt", sep="\t", row.names = FALSE, quote=FALSE)
```

## Intepret the theta results
```{r}
library("data.table")
library("GenomicRanges")
library("plyr")
theta <- fread("cache/theta.txt", data.table=FALSE)
names(theta)[1] <- "seq"
up5k <- read.table("cache/mt_gene_up5k.txt", header=TRUE)
names(up5k)
### define the subject file for theta values
grc <- with(theta, GRanges(seqnames=seq, IRanges(start=Pos, end=Pos)))
### define the query file for genomic feature
grf <- with(up5k, GRanges(seqnames=seq, IRanges(start=start, end=end), geneid=geneid))
### find overlaps between the two
tb <- findOverlaps(query=grf, subject=grc)
tb <- as.matrix(tb)
out1 <- as.data.frame(grf[tb[,1]])
out2 <- as.data.frame(grc[tb[,2]])
### for each genomic feature, find the sites with non-missing data
out <- cbind(out1, out2[, "start"]) 
names(out)[ncol(out)] <- "pos"
```
## Intepret the theta results

```{r}
#define unique identifier and merge with the thetas
out$uid <- paste(out$seqnames, out$pos, sep="_")
theta$uid <- paste(theta$seq, theta$Pos, sep="_")
df <- merge(out, theta[, c(-1, -2)], by="uid")
names(theta)
# for each upstream 5k region, how many theta values
mx <- ddply(df, .(geneid), summarise,
            Pairwise = mean(Pairwise, na.rm=TRUE),
            thetaH = mean(thetaH, na.rm=TRUE),
            nsites = length(uid))
```



```{r}
get_mean_theta <- function(gf_file="cache/mt_gene_up5k.txt"){
  # gf_file: gene feature file [chr, ="cache/mt_gene_up5k.txt"]
  
  theta <- fread("cache/theta.txt", data.table=FALSE)
  names(theta)[1] <- "seq"

  up5k <- read.table(gf_file, header=TRUE)

  ### define the subject file for theta values
  grc <- with(theta, GRanges(seqnames=seq, IRanges(start=Pos, end=Pos)))

  ### define the query file for genomic feature
  grf <- with(up5k, GRanges(seqnames=seq, IRanges(start=start, end=end), geneid=geneid))
    
  ### find overlaps between the two
  tb <- findOverlaps(query=grf, subject=grc)
  tb <- as.matrix(tb)
    
  out1 <- as.data.frame(grf[tb[,1]])
  out2 <- as.data.frame(grc[tb[,2]])
  ### for each genomic feature, find the sites with non-missing data
  out <- cbind(out1, out2[, "start"]) 
  names(out)[ncol(out)] <- "pos"

  #define unique identifier and merge with the thetas
  out$uid <- paste(out$seqnames, out$pos, sep="_")
  theta$uid <- paste(theta$seq, theta$Pos, sep="_")

  df <- merge(out, theta[, c(-1, -2)], by="uid")
  # for each upstream 5k region, how many theta values

  mx <- ddply(df, .(geneid), summarise,
            Pairwise = mean(Pairwise, na.rm=TRUE),
            thetaH = mean(thetaH, na.rm=TRUE),
            nsites = length(uid))
  return(mx)
}
```


```{r}
### apply the function
up5k <- get_mean_theta(gf_file="cache/mt_gene_up5k.txt")
down5k <- get_mean_theta(gf_file="cache/mt_gene_down5k.txt")
names(up5k)
```

```{r}
library("ggplot2")
up5k$feature <- "up 5k"
down5k$feature <- "down 5k"
res <- rbind(up5k, down5k)
names(up5k)
library(ggpubr)
compaired <- list(c("up 5k", "down 5k") )

#res <- as.data.frame(res)
ggplot(res, aes(x=feature, y=Pairwise, fill=feature)) + 
  geom_violin(trim=FALSE)+
  labs(title="Theta value", x="", y = "Log10 (theta)")+
  geom_boxplot(width=0.1, fill="white")+
  scale_fill_brewer(palette="Blues") + 
  geom_signif(comparisons = compaired,
              step_increase = 0.1,
              map_signif_level = F,
              test = t.test)+
  theme_classic()

library(ggpubr)
compaired <- list(c("up 5k", "down 5k")
                  )



```

